src/main/groovy/com/bmuschko/gradle/docker/tasks/container/DockerCopyFileFromContainer.groovy

```diff
@@ -15,55 +15,209 @@
 
-        try {
-            gzipOut = new GZIPOutputStream(outputStream)
-            gzipOut << input
+        // If user supplied an existing directory then we are responsible for naming and so
+        // will ensure file ends with '.tar'. If user supplied a regular file then use
+        // whichever name was passed in.
+        def fileName = new File(getRemotePath()).name
+        def compressedFileName = (hostDestination.exists() && hostDestination.isDirectory()) ?
+                (fileName.endsWith(".tar") ?: fileName + ".tar") :
+                hostDestination.name
+
+        def compressedFileLocation = (hostDestination.exists() && hostDestination.isDirectory()) ?
+                hostDestination :
+                hostDestination.parentFile
+
+        // If user supplied regular file ensure its parent location exists and if
+        // the regular file itself exists, delete to avoid clobbering.
+        if (hostDestination.exists()) {
+            if (!hostDestination.isDirectory())
+                hostDestination.delete()
+        } else {
+            hostDestination.parentFile.mkdirs()
         }
-        finally {
-            gzipOut?.close()
+
+        new File(compressedFileLocation,
+                compressedFileName).
+                withOutputStream { it << tarStream }
+
+    }
+
+    /**
+     * Copy regular file or directory from container to host
+     */
+    private void copyFile(InputStream tarStream, File hostDestination) {
+
+        def tempFile
+        def tempDir
+        try {
+
+            // Write tar to temp location since we are exploding it anyway
+            tempFile = File.createTempFile(UUID.randomUUID().toString(), ".tar")
+            tempFile.withOutputStream { it << tarStream }
+
+            // We are not allowed to rename tempDir's created in OS temp directory (as
+            // we do further downstream) which is why we are creating it relative to
+            // build directory
+            tempDir = new File("${project.buildDir}/${UUID.randomUUID().toString()}")
+            tempDir.mkdirs()
+            project.copy {
+                into tempDir
+                from project.tarTree(tempFile)
+            }
+            tempFile.delete()
+
+            /*
+                At this juncture we have 3 possibilities:
+
+                    1.) 0 files were found in which case we do nothing
+
+                    2.) 1 regular file was found
+
+                    3.) N regular files (and possibly directories) were found
+            */
+            def fileCount = 0
+            tempDir.eachFileRecurse(FileType.FILES) { fileCount++ }
+            if (fileCount == 0) {
+
+                logger.quiet "Nothing to copy."
+
+            } else if (fileCount == 1) {
+
+                // ensure regular file does not exist as we don't want clobbering
+                if (hostDestination.exists() && !hostDestination.isDirectory())
+                    hostDestination.delete()
```

Comment:
We'll need to make sure the return value of `delete()` and `mkdirs()` is handled. Otherwise, we'll swallow an error and the task will fail. Need to check for result and throw `GradleException` if not true.